### PR TITLE
Implement dependency resolution for services

### PR DESCRIPTION
**Pull request by swissspidy**
As taken from https://github.com/ampproject/amp-wp/pull/6484

Are there any dependencies we'd need to add already?

### CODE REVIEW

File: includes/Infrastructure/ServiceBasedPlugin.php [hunk 1]
```diff
@@ -396,7 +449,7 @@ function () use ( $class ) {
 	 * @param Injector $injector Injector instance to configure.
 	 * @return Injector Configured injector instance.
 	 */
-	protected function configure_injector( Injector $injector ) {
+	protected function configure_injector( Injector $injector ): Injector {
```
Comment (spacedmonkey) [inline includes/Infrastructure/ServiceBasedPlugin.php:592]: This made a problem for me with PHPstan as I recall. Why is this needed?

Comment (swissspidy) [reply includes/Infrastructure/ServiceBasedPlugin.php:592]: The method takes an Injector and returns an Injector, so it's reasonable to add a return type.

I don't see any PHPStan complaints here and tests pass, so seems all good.

**Comment by spacedmonkey** [comment]
I am not sure about adding this, until we need it. Is there any example service we could implement this on? 

How about `Jetpack` or `New_Relic`?

**Comment by spacedmonkey** [comment]
Silly question, how does this different from conditional service?

**Comment by swissspidy** [comment]
> Silly question, how does this different from conditional service?

This is for services that depend on another service, not services that depend on some outside condition (e.g. `is_admin()`).

For example if service A uses `Services::get( 'service_b' )` somewhere in its code, then B should be added as a dependency, so that it's guaranteed to be instantiated and exist first.

While we might not have obvious candidates for this in our code base yet (at least at first glance), it seems like a reasonable enhancement to facilitate further restructuring of the code base to use more services.

One opportunity there might be moving some of our functions in `namespace.php` to services implementing the `Activateable` and `Deactivateable` interfaces, where they could properly depend on the services they'd need.

Just some ideas though, haven't looked at it closely.

**Review by spacedmonkey — APPROVED** [review]

**Comment by swissspidy** [comment]
Sooo... this PR seems to uncover some flaws in our code.

For example, `Stories_Settings_Controller` relies on `Settings` first registering the settings, so that these are properly exposed in the REST API.

But with this PR it looks like `Settings` is not running yet when `Stories_Settings_Controller` is initialized. Adding the following to the controller class even causes a timeout:

```php
public static function get_requirements(): array {
return [
	'settings',
];
}
```

The timeout goes away if `Settings` does not implement `Delayed`. So... a bit confusing 🤔 

@pierlon do you perhaps have any thoughts or pointers?

**Comment by pierlon** [comment]
@swissspidy It seems you've uncovered a bug. With the example you stated above the `Settings` service wouldn't be available yet for `Stories_Settings_Controller` as its registration is delayed, and thus it won't be in the service container. Therefore `Stories_Settings_Controller` will be infinitely waiting on a service that will never exist in the service container until the `init` action has been triggered.

I'll see how best this can be resolved and test the changes over at https://github.com/ampproject/amp-wp/pull/6484.

**Comment by schlessera** [comment]
> Silly question, how does this different from conditional service?

The main purpose of this is to make the service registration independent of loading order. Relying on a specific order of things is always a latent bug, basically, so explicitly stating the requirements and adapting the loading order based on that is the only way to reliably deal with changing code and potentially even extending the service hierarchy.

**Comment by schlessera** [comment]
Regarding the `Settings` <=> `Stories_Settings_Controller` bug:
1. `Settings` is set up incorrectly. It is actually `Delayed` if you look at its `register()` method, but doesn't declare itself to be. So, `Settings` should be `Delayed`, with a `registration_action` of `'init'`.
2. The reordering implementation we currently have in the AMP plugin was the most simplistic implementation we could think of. However, it doesn't deal correctly with situations like the above. The interface and approach is still sound, but the actually reordering needs to happen in time, instead of only at the service array level. This means that when Service A depends on Service B, and Service B is delayed, then Service A needs to be Delayed to _after_ Service B (using an action). This will then work reliably in terms of timing.
However, keep in mind that this can still lead to issues. But these issues are design/dependency problems, not problems of the order resolution.

I'll adapt the PR in the AMP plugin to take delaying actions into account, and then we can adapt this one to ensure the problem is resolved.

**Comment by swissspidy** [comment]
> 1. `Settings` is set up incorrectly. It is actually `Delayed` if you look at its `register()` method, but doesn't declare itself to be. So, `Settings` should be `Delayed`, with a `registration_action` of `'init'`.

@schlessera `Settings` extends `Service_Base` which implements `Delayed`

**Comment by schlessera** [comment]
@swissspidy Yes, but `Settings`' registration action does nothing, except for further delaying...
Implement dependency resolution for services by swissspidy · Pull Request #  2021-08-02 at 12 27 14 PM

**Comment by swissspidy** [comment]
Oh, right, that doesn't make much sense 🤔   It's registered at init w/ prio 5 only to then call `register_settings()` at prio 10.

**Comment by schlessera** [comment]
Yep. The method `register_settings()` is actually what `Settings::register()` is supposed to be, which even the name already suggests.

**Comment by schlessera** [comment]
@swissspidy Is the main issue you've identified the first time around fixed now with the changes I made?

**Comment by swissspidy** [comment]
I believe so, yes! Thanks a lot!

Next I'll go through our services to properly identify & add dependencies where needed.

**Comment by spacedmonkey** [comment]
> Yep. The method register_settings() is actually what Settings::register() is supposed to be, which even the name already suggests.

Good catch @schlessera . This should never been this way. It is a simple fix. 

https://github.com/google/web-stories-wp/pull/8717

**Comment by swissspidy** [comment]
@schlessera I tried to implement https://github.com/ampproject/amp-wp/pull/6548 here, but there are lots of test failures because services are now only be registered once. Presumably I need to reset the ServiceContainer between tests to fix this... or somehow write better tests 🤔  Appreciate any pointers you might have

**Comment by schlessera** [comment]
@swissspidy I've had a brief look at the testing setup in this plugin. The errors and failures you have now point to a problem with how the tests are structured and executed.

What you have now is that the plugin is being directly loaded and triggered (hitting the `register()` method) via the PHPUnit bootstrapping logic, i.e. via `tests/phpunit/bootstrap.php`. That bootstrap file directly includes the bootstrap file of your plugin itself. This triggers the entire procedural execution chain of the plugin, because the actual call to the `$plugin->register()` method is not even hooked up to a WordPress action, but just executed right there in the include.

All of your tests just assume the entire plugin has been loaded and executed, none of them actual state what parts of the code are needed to be initialized. And this assumption happened to work because your plugin was loaded before even beginning with the tests. But it is loaded only a single time and never properly discarded.

Here's what I recommend, to get more reliable tests:
1. Hook the `$plugin->register()` up properly to the WordPress lifecycle. A good action is probably `plugins_loaded`. Doing anything directly when the plugin bootstrap file is included makes your plugin depend on the loading order and filesystem location, instead of the WP execution timing.
2. Create separate test cases for the contexts you want to test in and stick with these. As an example, for the AMP plugin we have a regular `TestCase` (for pure unit tests where we test a single class in isolation), the `WP_UnitTestCase` (for tests that make use of WordPress functionality), and the `DependencyInjectedTestCase` (which extends `WP_UnitTestCase`) and creates a fresh instance of the services infrastructure for each scenario. You can see that last one [here](https://github.com/ampproject/amp-wp/blob/2.1.3/tests/php/src/DependencyInjectedTestCase.php).

Note that 2. is the important one. By default, your plugin should not be fully loaded (once) before the test suite. Each test should start with a clean slate, decide via the test case what context to run within, and then make proper use of that context to do any missing arrangements before acting on the test.

I would bet that right now, you not only have tests that break due to the change I introduced in the service registration logic, but you'd probably have even more failures if you ran the tests in random order or in isolated processes. All failures that you can trigger in such a way are latent bugs and should be fixed (but it is often times only the test that is broken not the logic).

I hope this helps you makes sense of these failures. If not, ping me to discuss in more detail.

File: includes/AMP/Sanitization.php [hunk 2]
```diff
@@ -407,7 +404,7 @@ protected function get_sanitizers(): array {
 		$post = get_queried_object();
 
 		if ( $post instanceof \WP_Post && Story_Post_Type::POST_TYPE_SLUG === get_post_type( $post ) ) {
-			$video_cache_enabled = (bool) get_option( Settings::SETTING_NAME_VIDEO_CACHE );
+			$video_cache_enabled = (bool) Services::get( 'settings' )->get_setting( Settings::SETTING_NAME_VIDEO_CACHE );
```
Comment (spacedmonkey) [inline includes/AMP/Sanitization.php]: Why do it this way and not just inject a class instance?

Comment (spacedmonkey) [reply includes/AMP/Sanitization.php]: Also, doesn't this make the settings class a dependency of this one? Don't you need to register the setting before this point, to ensure the default values are set.

File: includes/Admin/Cross_Origin_Isolation.php [hunk 3]
```diff
@@ -81,6 +83,38 @@ public static function get_registration_action_priority(): int {
...
+		if ( ! $user_id ) {
+			return false;
+		}
+
+		// Cross-origin isolation is not needed if users can't upload files anyway.
+		if ( ! user_can( $user_id, 'upload_files' ) ) {
+			return false;
+		}
+
+		return rest_sanitize_boolean( Services::get( 'user_preferences' )->get_preference( $user_id, Preferences::MEDIA_OPTIMIZATION_META_KEY ) );
```
Comment (spacedmonkey) [inline includes/Admin/Cross_Origin_Isolation.php]: Same point about dependency injection here. Why not pass an instanse into the class.

File: includes/Tracking.php [hunk 4]
```diff
@@ -128,6 +129,17 @@ public static function get_registration_action(): string {
...
 
+	/**
+	 * Get the list of service IDs required for this service to be registered.
+	 *
+	 * @since 1.12.0
+	 *
+	 * @return string[] List of required services.
+	 */
+	public static function get_requirements(): array {
+		return [ 'user_preferences' ];
```
Comment (spacedmonkey) [inline includes/Tracking.php]: Isn't settings a requirement as well?

File: tests/phpunit/integration/tests/Admin/Cross_Origin_Isolation.php [hunk 5]
```diff
@@ -87,9 +92,8 @@ public function test_register() {
 	public function test_is_needed() {
 		wp_set_current_user( self::$admin_id );
 		update_user_meta( self::$admin_id, \Google\Web_Stories\User\Preferences::MEDIA_OPTIMIZATION_META_KEY, true );
-		$instance = new \Google\Web_Stories\Admin\Cross_Origin_Isolation();
-		$result   = $this->call_private_method( $instance, 'is_needed' );
-		$this->assertTrue( $result );
+
+		$this->assertTrue( \Google\Web_Stories\Admin\Cross_Origin_Isolation::is_needed() );
```
Comment (spacedmonkey) [inline tests/phpunit/integration/tests/Admin/Cross_Origin_Isolation.php:97]: Why can you do this?
```suggestion
		$this->assertTrue( $this->instance::is_needed() );
```

File: tests/phpunit/integration/tests/Experiments.php [hunk 6]
```diff
@@ -84,11 +84,14 @@ public function test_add_menu_page() {
 	 * @covers ::initialize_settings
 	 */
 	public function test_initialize_settings() {
+		global $wp_settings_fields, $wp_settings_sections;
+
 		$experiments = new \Google\Web_Stories\Experiments();
 		$experiments->initialize_settings();
 
-		$options = get_registered_settings();
```
Comment (spacedmonkey) [inline tests/phpunit/integration/tests/Experiments.php:86]: Why can this change?

File: tests/phpunit/integration/tests/User/Preferences.php [hunk 7]
```diff
@@ -208,7 +240,7 @@ public function test_enables_author_user_to_invalid_type() {
 		);
 
 		$response = rest_get_server()->dispatch( $request );
-		if ( is_a( $response, 'WP_REST_Response' ) ) {
+		if ( $response instanceof \WP_REST_Response ) {
```
Comment (spacedmonkey) [inline tests/phpunit/integration/tests/User/Preferences.php:230]: 👍

File: tests/phpunit/integration/tests/User/Preferences.php [hunk 8]
```diff
@@ -75,6 +93,9 @@ public function tearDown() {
 	 * @covers ::can_edit_current_user
 	 */
 	public function test_add_optin_field_to_rest_api() {
+		$this->instance->register();
+		( new Stories_Users_Controller() )->register();
```
Comment (spacedmonkey) [inline tests/phpunit/integration/tests/User/Preferences.php]: Why not have another class property called `$controller`, stop this repeated code.

File: includes/Infrastructure/ServiceBasedPlugin.php [hunk 9]
```diff
@@ -366,7 +463,37 @@ protected function get_identifier_from_fqcn( $fqcn ): string {
...
+	 *
+	 * @param string                       $id ID of the service to register.
+	 * @param HasRequirements|class-string $class Class of the service to register.
+	 * @return void
+	 */
+	protected function schedule_potential_service_registration( $id, $class ) {
+		if ( is_a( $class, Delayed::class, true ) ) {
+			$registration_action = $class::get_registration_action();
+
+			if ( \did_action( $registration_action ) ) {
```
Comment (spacedmonkey) [inline includes/Infrastructure/ServiceBasedPlugin.php]: Don't need the slash here.

File: includes/Infrastructure/ServiceBasedPlugin.php [hunk 10]
```diff
@@ -366,7 +463,37 @@ protected function get_identifier_from_fqcn( $fqcn ): string {
...
+				$this->maybe_register_service( $id, $class );
+			} else {
+				\add_action(
```
Comment (spacedmonkey) [inline includes/Infrastructure/ServiceBasedPlugin.php]: Should also get rid of slash here as well.

File: tests/phpunit/integration/tests/Admin/Cross_Origin_Isolation.php (hunk 5 above)
Comment (swissspidy) [reply tests/phpunit/integration/tests/Admin/Cross_Origin_Isolation.php:97]: Because it's a static method and should be called as such ideally.

File: tests/phpunit/integration/tests/Experiments.php (hunk 6 above)
Comment (swissspidy) [reply tests/phpunit/integration/tests/Experiments.php:86]: `get_registered_settings` checks which settings are registered.

Setting registration is the job of the `Settings` class, not the `Experiments` class.

Here we are only concerned about testing the `Experiments` class though.

File: includes/AMP/Sanitization.php [hunk 11]
```diff
@@ -407,7 +424,7 @@ protected function get_sanitizers(): array {
 		$post = get_queried_object();
 
 		if ( $post instanceof \WP_Post && Story_Post_Type::POST_TYPE_SLUG === get_post_type( $post ) ) {
-			$video_cache_enabled = (bool) get_option( Settings::SETTING_NAME_VIDEO_CACHE );
+			$video_cache_enabled = (bool) $this->settings->get_setting( Settings::SETTING_NAME_VIDEO_CACHE );
```
Comment (spacedmonkey) [inline includes/AMP/Sanitization.php]:
```suggestion
			$video_cache_enabled = (bool) $this->settings->get_setting( $this->settings::SETTING_NAME_VIDEO_CACHE );
```

Would this work?

File: includes/Admin/Cross_Origin_Isolation.php (hunk 3 above)
Comment (spacedmonkey) [reply includes/Admin/Cross_Origin_Isolation.php]: This line feel a little overly long. How about something like this. 
```suggestion
        $user_preferences = Services::get( 'user_preferences' );
        
		return rest_sanitize_boolean( $user_preferences->get_preference( $user_id, $user_preferences::MEDIA_OPTIMIZATION_META_KEY ) );
```

File: tests/phpunit/integration/tests/Media/Media_Source_Taxonomy.php [hunk 12]
```diff
@@ -102,9 +112,9 @@ public function test_rest_api_init() {
 		);
 
 		set_post_thumbnail( $video_attachment_id, $poster_attachment_id );
-		wp_set_object_terms( $video_attachment_id, 'editor', $media_source->get_taxonomy_slug() );
+		wp_set_object_terms( $video_attachment_id, 'editor', $this->instance->get_taxonomy_slug() );
 
-		$request  = new WP_REST_Request( \WP_REST_Server::READABLE, sprintf( '/web-stories/v1/media/%d', $video_attachment_id ) );
+		$request  = new WP_REST_Request( \WP_REST_Server::READABLE, sprintf( '/wp/v2/media/%d', $video_attachment_id ) );
```
Comment (spacedmonkey) [inline tests/phpunit/integration/tests/Media/Media_Source_Taxonomy.php:117]: Why was this changed?

File: tests/phpunit/integration/tests/Widgets/Stories.php [hunk 13]
```diff
@@ -28,50 +27,52 @@
...
 	 * @return void
 	 */
-	public static function set_up_before_class() {
-		$assets       = new \Google\Web_Stories\Assets();
-		self::$testee = new Testee( $assets );
+	public function set_up() {
+		parent::set_up();
+
+		$assets         = new \Google\Web_Stories\Assets();
+		$this->instance = new \Google\Web_Stories\Widgets\Stories( $assets );
```
Comment (spacedmonkey) [inline tests/phpunit/integration/tests/Widgets/Stories.php:45]: Can't we use `DependencyInjectedTestCase` here?

File: tests/phpunit/integration/tests/User/Preferences.php [hunk 14]
```diff
@@ -53,15 +61,21 @@ public static function wpSetUpBeforeClass( $factory ) {
 	}
 
 	public function set_up() {
-		$up = new \Google\Web_Stories\User\Preferences();
-		$up->register();
 		parent::set_up();
+
+		$this->instance = new \Google\Web_Stories\User\Preferences();
```
Comment (spacedmonkey) [inline tests/phpunit/integration/tests/User/Preferences.php]: Can't we use `DependencyInjectedTestCase` here?

File: tests/phpunit/integration/tests/User/Preferences.php [hunk 15]
```diff
@@ -53,15 +61,21 @@ public static function wpSetUpBeforeClass( $factory ) {
...
 
 	public function tear_down() {
+		delete_user_meta( self::$admin_id, \Google\Web_Stories\User\Preferences::MEDIA_OPTIMIZATION_META_KEY );
+		delete_user_meta( self::$admin_id, \Google\Web_Stories\User\Preferences::OPTIN_META_KEY );
+		delete_user_meta( self::$author_id, \Google\Web_Stories\User\Preferences::MEDIA_OPTIMIZATION_META_KEY );
+		delete_user_meta( self::$author_id, \Google\Web_Stories\User\Preferences::OPTIN_META_KEY );
+
 		unregister_meta_key( 'user', \Google\Web_Stories\User\Preferences::OPTIN_META_KEY );
 		unregister_meta_key( 'user', \Google\Web_Stories\User\Preferences::ONBOARDING_META_KEY );
 		unregister_meta_key( 'user', \Google\Web_Stories\User\Preferences::MEDIA_OPTIMIZATION_META_KEY );
```
Comment (spacedmonkey) [inline tests/phpunit/integration/tests/User/Preferences.php:78]:
```suggestion
		delete_user_meta( self::$admin_id, $this->instance::MEDIA_OPTIMIZATION_META_KEY );
		delete_user_meta( self::$admin_id, $this->instance::OPTIN_META_KEY );
		delete_user_meta( self::$author_id, $this->instance::MEDIA_OPTIMIZATION_META_KEY );
		delete_user_meta( self::$author_id, $this->instance::OPTIN_META_KEY );

		unregister_meta_key( 'user',$this->instance::OPTIN_META_KEY );
		unregister_meta_key( 'user',$this->instance::ONBOARDING_META_KEY );
		unregister_meta_key( 'user', $this->instance::MEDIA_OPTIMIZATION_META_KEY );
```

Does this make sense?

File: tests/phpunit/integration/tests/User/Preferences.php [hunk 16]
```diff
@@ -70,12 +84,14 @@ public function tear_down() {
...
 	public function test_add_optin_field_to_rest_api() {
+		$this->instance->register();
+
 		wp_set_current_user( self::$admin_id );
 
 		add_user_meta( self::$admin_id, \Google\Web_Stories\User\Preferences::OPTIN_META_KEY, true );
 		add_user_meta( self::$admin_id, \Google\Web_Stories\User\Preferences::MEDIA_OPTIMIZATION_META_KEY, true );
 
-		$request  = new WP_REST_Request( \WP_REST_Server::READABLE, sprintf( '/web-stories/v1/users/%d', self::$admin_id ) );
+		$request  = new WP_REST_Request( \WP_REST_Server::READABLE, sprintf( '/wp/v2/users/%d', self::$admin_id ) );
```
Comment (spacedmonkey) [inline tests/phpunit/integration/tests/User/Preferences.php:95]: Why change this?

File: tests/phpunit/integration/tests/REST_API/Stories_Users_Controller.php [hunk 17]
```diff
@@ -50,13 +58,28 @@ public static function wpSetUpBeforeClass( $factory ) {
...
 			]
 		);
+	}
+
+	public function set_up() {
+		parent::set_up();
+
+		$this->controller = new \Google\Web_Stories\REST_API\Stories_Users_Controller();
+
+		$this->set_permalink_structure( '/%postname%/' );
```
Comment (spacedmonkey) [inline tests/phpunit/integration/tests/REST_API/Stories_Users_Controller.php]: Do we still need this?

File: tests/phpunit/integration/tests/REST_API/Stories_Media_Controller.php [hunk 18]
```diff
@@ -73,10 +80,26 @@ public static function wpSetUpBeforeClass( $factory ) {
...
+	public function set_up() {
+		parent::set_up();
+
+		$this->controller = new \Google\Web_Stories\REST_API\Stories_Media_Controller();
+
+		$this->set_permalink_structure( '/%postname%/' );
+	}
+
+	public function tear_down() {
+		$this->set_permalink_structure( '' );
```
Comment (spacedmonkey) [inline tests/phpunit/integration/tests/REST_API/Stories_Media_Controller.php]: It's this already part of the tear_down?

File: tests/phpunit/integration/tests/REST_API/Page_Template_Controller.php [hunk 19]
```diff
@@ -115,6 +122,20 @@ public static function wpSetUpBeforeClass( $factory ) {
 		);
 	}
 
+	public function set_up() {
+		parent::set_up();
+
+		$this->controller = new \Google\Web_Stories\REST_API\Page_Template_Controller( \Google\Web_Stories\Story_Post_Type::POST_TYPE_SLUG );
```
Comment (spacedmonkey) [inline tests/phpunit/integration/tests/REST_API/Page_Template_Controller.php]: Post type seems wrong here.

**Review by spacedmonkey — CHANGES_REQUESTED** [review]
Some changes required

**Comment by spacedmonkey** [comment]
For any class that the injects settings/user_perfernce classes, don't these also need to be service dependency? To ensure that the user meta / settings is registered?

File: includes/Tracking.php [hunk 20]
```diff
@@ -85,11 +100,21 @@ class Tracking extends Service_Base {
 	 * @param Experiments $experiments Experiments instance.
 	 * @param Site_Kit    $site_kit Site_Kit instance.
 	 * @param Assets      $assets Assets instance.
+	 * @param Settings    $settings Settings instance.
+	 * @param Preferences $preferences Preferences instance.
 	 */
-	public function __construct( Experiments $experiments, Site_Kit $site_kit, Assets $assets ) {
+	public function __construct(
```
Comment (spacedmonkey) [inline includes/Tracking.php:106]: This class needs to have Settings and Preferences classes as dependencies.

File: tests/phpunit/integration/includes/TestCase.php [hunk 21]
```diff
@@ -45,8 +45,6 @@ public function tear_down() {
 			$_SERVER['HTTPS']
 		);
 
-		$this->clean_up_global_scope();
```
Comment (spacedmonkey) [inline tests/phpunit/integration/includes/TestCase.php:48]: Why remove this?

File: tests/phpunit/integration/tests/AMP/Sanitization.php [hunk 22]
```diff
@@ -64,7 +64,7 @@ public function test_scripts_get_moved_to_head() {
 		<?php
 		$original_html = ob_get_clean();
 
-		$sanitization = new \Google\Web_Stories\AMP\Sanitization();
+		$sanitization = $this->injector->make( \Google\Web_Stories\AMP\Sanitization::class );
```
Comment (spacedmonkey) [inline tests/phpunit/integration/tests/AMP/Sanitization.php]: Shouldn't this be in a set_up method?

File: tests/phpunit/integration/tests/AdSense.php [hunk 23]
```diff
@@ -42,7 +42,7 @@ public function tear_down() {
 	 * @covers ::register
 	 */
 	public function test_register() {
-		$adsense = new \Google\Web_Stories\AdSense();
+		$adsense = $this->injector->make( \Google\Web_Stories\AdSense::class );
```
Comment (spacedmonkey) [inline tests/phpunit/integration/tests/AdSense.php]: Should this be part of the set_up?

File: tests/phpunit/integration/tests/Ad_Manager.php [hunk 24]
```diff
@@ -40,7 +40,7 @@ public function tear_down() {
 	 * @covers ::register
 	 */
 	public function test_register() {
-		$adsense = new \Google\Web_Stories\Ad_Manager();
+		$adsense = $this->injector->make( \Google\Web_Stories\Ad_Manager::class );
```
Comment (spacedmonkey) [inline tests/phpunit/integration/tests/Ad_Manager.php]: Use an instance and setup in set_up method?

File: tests/phpunit/integration/tests/Admin/Cross_Origin_Isolation.php [hunk 25]
```diff
@@ -17,141 +17,142 @@
...
 		);
+
+		$this->instance = $this->injector->make( \Google\Web_Stories\Admin\Cross_Origin_Isolation::class );
 	}
 
-	public function set_up() {
-		parent::set_up();
+	public function tear_down() {
+		delete_user_meta( $this->admin_id, \Google\Web_Stories\User\Preferences::MEDIA_OPTIMIZATION_META_KEY );
+		delete_user_meta( $this->contributor_id, \Google\Web_Stories\User\Preferences::MEDIA_OPTIMIZATION_META_KEY );
```
Comment (spacedmonkey) [inline tests/phpunit/integration/tests/Admin/Cross_Origin_Isolation.php:66]:
```suggestion
		delete_user_meta( $this->contributor_id, \Google\Web_Stories\User\Preferences::MEDIA_OPTIMIZATION_META_KEY );
		delete_user_meta( $this->contributor_id, \Google\Web_Stories\User\Preferences::ONBOARDING_META_KEY );
```

File: tests/phpunit/integration/tests/Analytics.php [hunk 26]
```diff
@@ -22,13 +22,12 @@
...
  */
-class Analytics extends TestCase {
-
+class Analytics extends DependencyInjectedTestCase {
 	/**
 	 * @covers ::register
 	 */
 	public function test_register() {
-		$analytics = new \Google\Web_Stories\Analytics( new \Google\Web_Stories\Experiments() );
+		$analytics = $this->injector->make( \Google\Web_Stories\Analytics::class );
```
Comment (spacedmonkey) [inline tests/phpunit/integration/tests/Analytics.php]: Instance and set_up method?

File: tests/phpunit/integration/tests/Media/Video/Poster.php [hunk 27]
```diff
@@ -278,18 +281,16 @@ public function test_is_poster() {
...
-		$media_source         = new Media_Source_Taxonomy();
-		$poster               = new \Google\Web_Stories\Media\Video\Poster( $media_source );
 
-		$result1 = $this->call_private_method( $poster, 'is_poster', [ $poster_attachment_id ] );
+		$result1 = $this->call_private_method( $this->instance, 'is_poster', [ $poster_attachment_id ] );
 		$this->assertFalse( $result1 );
 
-		wp_set_object_terms( $poster_attachment_id, 'editor', $media_source->get_taxonomy_slug() );
-		$result2 = $this->call_private_method( $poster, 'is_poster', [ $poster_attachment_id ] );
+		wp_set_object_terms( $poster_attachment_id, 'editor', Services::get( 'media.media_source' )->get_taxonomy_slug() );
```
Comment (spacedmonkey) [inline tests/phpunit/integration/tests/Media/Video/Poster.php]: Let's use the class instance here. 

So
```
$media_source         = $this->injector->make( \Google\Web_Stories\Media\Media_Source_Taxonomy::class );
```

File: includes/AMP/Sanitization.php (hunk 11 above)
Comment (swissspidy) [reply includes/AMP/Sanitization.php]: It would. Will add this.

File: tests/phpunit/integration/tests/Media/Media_Source_Taxonomy.php (hunk 12 above)
Comment (swissspidy) [reply tests/phpunit/integration/tests/Media/Media_Source_Taxonomy.php:117]: As discussed today, this test only needs to test that the field is added to the REST API, not that our custom endpoints are available.

File: tests/phpunit/integration/tests/Widgets/Stories.php (hunk 13 above)
Comment (swissspidy) [reply tests/phpunit/integration/tests/Widgets/Stories.php:45]: The `Stories` class itself is not a service, so that's not possible.

File: tests/phpunit/integration/tests/User/Preferences.php (hunk 16 above)
Comment (swissspidy) [reply tests/phpunit/integration/tests/User/Preferences.php:95]: https://github.com/google/web-stories-wp/pull/8513#discussion_r721332879

File: tests/phpunit/integration/includes/TestCase.php (hunk 21 above)
Comment (swissspidy) [reply tests/phpunit/integration/includes/TestCase.php:48]: `\WP_UnitTestCase_Base::set_up()` already does this for us, so this is unnecessary here

File: tests/phpunit/integration/tests/Admin/Cross_Origin_Isolation.php (hunk 25 above)
Comment (swissspidy) [reply tests/phpunit/integration/tests/Admin/Cross_Origin_Isolation.php:66]: Why would this be needed? Neither the `Cross_Origin_Isolation` class nor this test use this user meta

**Comment by swissspidy** [comment]
> For any class that the injects settings/user_perfernce classes, don't these also need to be service dependency? To ensure that the user meta / settings is registered?

Nope, the `get_requirements()` method to add requirements is only needed for services that reference another service in the static `is_needed()` class.

So far `Cross_Origin_Isolation` is the only place where this is needed.

When an instance is injected via the constructor, `ServiceBasedPlugin` and `SimpleInjector` will create those instances first.

This is in line with how it works in the AMP plugin:

https://github.com/ampproject/amp-wp/blob/543aa27087c6fe1847746eda84facea44a5a90bf/src/Admin/ValidationCounts.php#L42-L61
https://github.com/ampproject/amp-wp/blob/543aa27087c6fe1847746eda84facea44a5a90bf/src/Admin/Polyfills.php#L38-L56
https://github.com/ampproject/amp-wp/blob/d1652901e9eb0316102fe8e1256f73d30d2c866e/src/Admin/PairedBrowsing.php#L69-L103

File: tests/phpunit/integration/tests/User/Preferences.php (hunk 15 above)
Comment (swissspidy) [reply tests/phpunit/integration/tests/User/Preferences.php:78]: I don't think we really need to use the instance here for the test cleanup, but I'm indifferent.

File: tests/phpunit/integration/tests/REST_API/Stories_Users_Controller.php (hunk 17 above)
Comment (spacedmonkey) [reply tests/phpunit/integration/tests/REST_API/Stories_Users_Controller.php]: This is done here https://github.com/google/web-stories-wp/blob/29a4bd1ef802e40059cfe16e8e77ff80ebde7854/tests/phpunit/integration/includes/Test_REST_TestCase.php#L53

File: tests/phpunit/integration/tests/REST_API/Stories_Media_Controller.php [hunk 28]
```diff
@@ -73,10 +80,20 @@ public static function wpSetUpBeforeClass( $factory ) {
 		);
 	}
 
+	public function set_up() {
+		parent::set_up();
+
+		$this->controller = new \Google\Web_Stories\REST_API\Stories_Media_Controller();
+
+		$this->set_permalink_structure( '/%postname%/' );
```
Comment (spacedmonkey) [inline tests/phpunit/integration/tests/REST_API/Stories_Media_Controller.php]: See https://github.com/google/web-stories-wp/pull/8513/files#r722067897

File: tests/phpunit/integration/tests/REST_API/Stories_Autosaves_Controller.php [hunk 29]
```diff
@@ -26,13 +27,24 @@
  *
  * @package Google\Web_Stories\Tests\REST_API
  *
- * @coversDefaultClass \Google\Web_Stories\REST_API\Autosaves_Controller
+ * @coversDefaultClass Controller
```
Comment (spacedmonkey) [inline tests/phpunit/integration/tests/REST_API/Stories_Autosaves_Controller.php]: Is this correct?

**Comment by spacedmonkey** [comment]
There is any value in adding settings / preference classes to share instances?

https://github.com/google/web-stories-wp/blob/29a4bd1ef802e40059cfe16e8e77ff80ebde7854/includes/Plugin.php#L164-L176

File: includes/KSES.php [hunk 30]
```diff
@@ -86,17 +97,25 @@ public static function get_registration_action_priority(): int {
...
 				'revision' === $data['post_type'] &&
 				! empty( $data['post_parent'] ) &&
-				Story_Post_Type::POST_TYPE_SLUG === get_post_type( $data['post_parent'] )
+				get_post_type( $data['post_parent'] ) === $this->story_post_type::POST_TYPE_SLUG
 			)
 		) {
 			return $data;
 		}
 
+		if ( ! $this->get_post_type_cap( $this->story_post_type::POST_TYPE_SLUG, 'edit_posts' ) ) {
```
Comment (spacedmonkey) [inline includes/KSES.php:115]: We do this now and remove the trait on this class.
```suggestion
		if ( ! $this->story_post_type->get_post_type_cap( $this->story_post_type::POST_TYPE_SLUG, 'edit_posts' ) ) {
```

File: tests/phpunit/integration/tests/REST_API/Stories_Media_Controller.php (hunk 28 above)
Comment (swissspidy) [reply tests/phpunit/integration/tests/REST_API/Stories_Media_Controller.php]: I'll remove, but why are we defaulting to pretty permalinks in all tests? 🤔

File: includes/KSES.php (hunk 30 above)
Comment (swissspidy) [reply includes/KSES.php:115]: We should probably get rid of this trait eventually when we can just access the instance..

Comment (swissspidy) [reply includes/KSES.php:115]: Also, I can't do this change because `get_post_type_cap` is `protected`

File: includes/Page_Template_Post_Type.php [hunk 31]
```diff
@@ -50,8 +69,8 @@ class Page_Template_Post_Type extends Service_Base {
 	 * @return void
 	 */
 	public function register() {
-		$edit_posts   = $this->get_post_type_cap_name( Story_Post_Type::POST_TYPE_SLUG, 'edit_posts' );
-		$delete_posts = $this->get_post_type_cap_name( Story_Post_Type::POST_TYPE_SLUG, 'delete_posts' );
+		$edit_posts   = $this->get_post_type_cap_name( $this->story_post_type::POST_TYPE_SLUG, 'edit_posts' );
+		$delete_posts = $this->get_post_type_cap_name( $this->story_post_type::POST_TYPE_SLUG, 'delete_posts' );
```
Comment (spacedmonkey) [inline includes/Page_Template_Post_Type.php:73]:
```suggestion
		$edit_posts   = $this->story_post_type->get_post_type_cap_name( $this->story_post_type::POST_TYPE_SLUG, 'edit_posts' );
		$delete_posts = $this->story_post_type->get_post_type_cap_name( $this->story_post_type::POST_TYPE_SLUG, 'delete_posts' );
```

Comment (swissspidy) [reply includes/Page_Template_Post_Type.php:73]: This does not work. `get_post_type_cap_name` is `protected`. If we wanna change this, then we can probably just get rid of that trait too. But in a separate PR.

Comment (spacedmonkey) [reply includes/Page_Template_Post_Type.php:73]: I was wonder about changing the Post Type trait an abstract base class for all post types. We can make post types a little more generic. When we can pass a   `story_post_type` object into classes that need to be capability checks.

File: includes/KSES.php (hunk 30 above)
Comment (spacedmonkey) [reply includes/KSES.php:115]: Let's do this in a follow up PR.

**Comment by spacedmonkey** [comment]
There are still a couple of instances of `update_option` and `get_option` in the code, do we need to fix these as well?

File: includes/Admin/Cross_Origin_Isolation.php (hunk 3 above)
Comment (swissspidy) [reply includes/Admin/Cross_Origin_Isolation.php]: `is_needed()` is a **static** method. Can't use injection there.

Will split the function call across multiple lines.

**Review by spacedmonkey — APPROVED** [review]
Great work! Let's ship it :shipit:
